File: packages/sdk-python/agenteconomy/utils/datum.py
"""Plutus Data encoding/decoding for all datum types."""

from __future__ import annotations

from typing import Any

from agenteconomy.types import (
    BountyDatum, WalletPolicyDatum, ReputationDatum,
    BountyStatus, BountyStatusType, VerificationType,
    BountyCategory, Difficulty, AssetId,
)

# Plutus Data representation as nested dicts/lists
PlutusData = dict[str, Any] | int | str | list | None


def constr(index: int, fields: list[Any]) -> dict[str, Any]:
    """Create a Plutus Data constructor."""
    return {"constructor": index, "fields": fields}


def option_to_data(val: Any, encoder: Any = None) -> dict[str, Any]:
    """Encode an optional value as Plutus None/Some."""
    if val is None:
        return constr(1, [])
    encoded = encoder(val) if encoder else val
    return constr(0, [encoded])


def data_to_option(data: Any, decoder: Any = None) -> Any:
    """Decode Plutus None/Some to Python optional."""
    if isinstance(data, dict) and "constructor" in data:
        if data["constructor"] == 1:
            return None
        if data["constructor"] == 0 and data["fields"]:
            return decoder(data["fields"][0]) if decoder else data["fields"][0]
    return None


CATEGORY_INDEX = {
    BountyCategory.DATA_EXTRACTION: 0,
    BountyCategory.CODE_GEN: 1,
    BountyCategory.RESEARCH: 2,
    BountyCategory.CONTENT: 3,
    BountyCategory.ON_CHAIN: 4,
    BountyCategory.TRANSLATION: 5,
    BountyCategory.MODERATION: 6,
}

INDEX_CATEGORY = {v: k for k, v in CATEGORY_INDEX.items()}

DIFFICULTY_INDEX = {
    Difficulty.EASY: 0,
    Difficulty.MEDIUM: 1,
    Difficulty.HARD: 2,
    Difficulty.EXPERT: 3,
}

INDEX_DIFFICULTY = {v: k for k, v in DIFFICULTY_INDEX.items()}


def encode_bounty_datum(datum: BountyDatum) -> dict[str, Any]:
    """Encode BountyDatum to Plutus Data."""
    return constr(0, [
        datum.bounty_id,
        datum.version,
        datum.poster,
        datum.title,
        datum.description_ipfs,
        constr(CATEGORY_INDEX[datum.category], []),
        constr(DIFFICULTY_INDEX[datum.difficulty], []),
        datum.reward_lovelace,
        option_to_data(datum.reward_token, lambda t: constr(0, [t.policy_id, t.asset_name])),
        datum.deposit_lovelace,
        datum.deadline,
        datum.claim_window_ms,
        _encode_verification(datum.verification),
        datum.dispute_window_ms,
        option_to_data(datum.result_schema_ipfs),
        option_to_data(datum.allowed_agents),
        option_to_data(datum.min_reputation_score),
        _encode_status(datum.status),
        datum.created_at,
        datum.metadata_hash,
    ])


def decode_bounty_datum(data: dict[str, Any]) -> BountyDatum:
    """Decode Plutus Data to BountyDatum."""
    f = data["fields"]
    return BountyDatum(
        bounty_id=f[0],
        version=f[1],
        poster=f[2],
        title=f[3],
        description_ipfs=f[4],
        category=INDEX_CATEGORY[f[5]["constructor"]],
        difficulty=INDEX_DIFFICULTY[f[6]["constructor"]],
        reward_lovelace=f[7],
        reward_token=data_to_option(f[8], lambda d: AssetId(d["fields"][0], d["fields"][1])),
        deposit_lovelace=f[9],
        deadline=f[10],
        claim_window_ms=f[11],
        verification=_decode_verification(f[12]),
        dispute_window_ms=f[13],
        result_schema_ipfs=data_to_option(f[14]),
        allowed_agents=data_to_option(f[15]),
        min_reputation_score=data_to_option(f[16]),
        status=_decode_status(f[17]),
        created_at=f[18],
        metadata_hash=f[19],
    )


def encode_wallet_policy_datum(datum: WalletPolicyDatum) -> dict[str, Any]:
    """Encode WalletPolicyDatum to Plutus Data."""
    return constr(0, [
        datum.owner_pkh,
        datum.agent_pkh,
        datum.daily_limit_lovelace,
        datum.per_tx_limit_lovelace,
        datum.whitelisted_scripts,
        datum.whitelisted_addresses,
        datum.require_owner_above,
        [constr(0, [t.policy_id, t.asset_name]) for t in datum.allowed_tokens],
        datum.max_fee_lovelace,
        option_to_data(datum.pause_until),
        option_to_data(
            datum.multisig_threshold,
            lambda ms: constr(0, [ms[0], ms[1]]),
        ),
        datum.nonce,
        datum.policy_version,
    ])


def decode_wallet_policy_datum(data: dict[str, Any]) -> WalletPolicyDatum:
    """Decode Plutus Data to WalletPolicyDatum."""
    f = data["fields"]
    return WalletPolicyDatum(
        owner_pkh=f[0],
        agent_pkh=f[1],
        daily_limit_lovelace=f[2],
        per_tx_limit_lovelace=f[3],
        whitelisted_scripts=f[4],
        whitelisted_addresses=f[5],
        require_owner_above=f[6],
        allowed_tokens=[AssetId(t["fields"][0], t["fields"][1]) for t in f[7]],
        max_fee_lovelace=f[8],
        pause_until=data_to_option(f[9]),
        multisig_threshold=data_to_option(f[10], lambda d: (d["fields"][0], d["fields"][1])),
        nonce=f[11],
        policy_version=f[12],
    )


def encode_reputation_datum(datum: ReputationDatum) -> dict[str, Any]:
    """Encode ReputationDatum to Plutus Data."""
    return constr(0, [
        datum.agent_pkh,
        datum.total_completed,
        datum.total_earned_lovelace,
        datum.total_disputed,
        datum.success_rate_bps,
        datum.avg_completion_ms,
        [[cat_id, score] for cat_id, score in datum.category_scores],
        datum.last_active,
        datum.badge_tokens,
        datum.registered_at,
    ])


def decode_reputation_datum(data: dict[str, Any]) -> ReputationDatum:
    """Decode Plutus Data to ReputationDatum."""
    f = data["fields"]
    return ReputationDatum(
        agent_pkh=f[0],
        total_completed=f[1],
        total_earned_lovelace=f[2],
        total_disputed=f[3],
        success_rate_bps=f[4],
        avg_completion_ms=f[5],
        category_scores=[(pair[0], pair[1]) for pair in f[6]],
        last_active=f[7],
        badge_tokens=f[8],
        registered_at=f[9],
    )


def _encode_verification(v: VerificationType) -> dict[str, Any]:
    if v.type == "Optimistic":
        return constr(0, [v.dispute_window_ms])
    elif v.type == "HumanReview":
        return constr(1, [])
    elif v.type == "OracleSigned":
        return constr(2, [v.oracle_pub_key])
    else:
        return constr(3, [])


def _decode_verification(data: dict[str, Any]) -> VerificationType:
    idx = data["constructor"]
    fields = data["fields"]
    if idx == 0:
        return VerificationType(type="Optimistic", dispute_window_ms=fields[0])
    elif idx == 1:
        return VerificationType(type="HumanReview")
    elif idx == 2:
        return VerificationType(type="OracleSigned", oracle_pub_key=fields[0])
    else:
        return VerificationType(type="JudgeAgent")


def _encode_status(s: BountyStatus) -> dict[str, Any]:
    status_map = {
        BountyStatusType.OPEN: (0, []),
        BountyStatusType.CLAIMED: (1, [s.agent, s.claimed_at]),
        BountyStatusType.WORK_SUBMITTED: (2, [s.result_ipfs, s.submitted_at]),
        BountyStatusType.COMPLETED: (3, [s.result_ipfs, s.completed_at]),
        BountyStatusType.DISPUTED: (4, []),
        BountyStatusType.REFUNDED: (5, []),
    }
    idx, fields = status_map[s.type]
    return constr(idx, fields)


def _decode_status(data: dict[str, Any]) -> BountyStatus:
    idx = data["constructor"]
    fields = data["fields"]
    if idx == 0:
        return BountyStatus(type=BountyStatusType.OPEN)
    elif idx == 1:
        return BountyStatus(type=BountyStatusType.CLAIMED, agent=fields[0], claimed_at=fields[1])
    elif idx == 2:
        return BountyStatus(type=BountyStatusType.WORK_SUBMITTED, result_ipfs=fields[0], submitted_at=fields[1])
    elif idx == 3:
        return BountyStatus(type=BountyStatusType.COMPLETED, result_ipfs=fields[0], completed_at=fields[1])
    elif idx == 4:
        return BountyStatus(type=BountyStatusType.DISPUTED)
    else:
        return BountyStatus(type=BountyStatusType.REFUNDED)

File: packages/sdk-ts/src/utils/datum.ts
// ─────────────────────────────────────────────────────────
// Plutus Data encoding/decoding for all datum types
// ─────────────────────────────────────────────────────────

import type {
  BountyDatum,
  BountyRedeemer,
  BountyStatus,
  VerificationType,
  WalletPolicyDatum,
  ReputationDatum,
  BountyCategory,
  Difficulty,
} from "../types.js";

export interface Constr {
  constructor: number;
  fields: PlutusData[];
}

// Re-export a type alias for Plutus Data representation
export type PlutusData =
  | Constr
  | number
  | bigint
  | string
  | PlutusData[]
  | Map<PlutusData, PlutusData>;

function constr(index: number, fields: PlutusData[]): Constr {
  return { constructor: index, fields };
}

function isConstr(data: PlutusData): data is Constr {
  return typeof data === "object" && data !== null && !Array.isArray(data) && !(data instanceof Map) && "constructor" in data;
}

function optionToData(val: unknown, encoder?: (v: unknown) => PlutusData): Constr {
  if (val === null || val === undefined) {
    return constr(1, []); // None
  }
  const encoded = encoder ? encoder(val) : (val as PlutusData);
  return constr(0, [encoded]); // Some(val)
}

function dataToOption<T>(data: PlutusData, decoder: (d: PlutusData) => T): T | null {
  if (isConstr(data)) {
    if (data.constructor === 1) return null; // None
    if (data.constructor === 0 && data.fields.length > 0) {
      return decoder(data.fields[0]!);
    }
  }
  return null;
}

// ── Category encoding ───────────────────────────────────

const CATEGORY_INDEX: Record<BountyCategory, number> = {
  DataExtraction: 0,
  CodeGen: 1,
  Research: 2,
  Content: 3,
  OnChain: 4,
  Translation: 5,
  Moderation: 6,
};

const INDEX_CATEGORY: BountyCategory[] = [
  "DataExtraction", "CodeGen", "Research", "Content",
  "OnChain", "Translation", "Moderation",
];

function encodeCategory(cat: BountyCategory): PlutusData {
  return constr(CATEGORY_INDEX[cat], []);
}

function decodeCategory(data: PlutusData): BountyCategory {
  if (isConstr(data)) {
    const cat = INDEX_CATEGORY[data.constructor];
    if (cat) return cat;
  }
  throw new Error(`Invalid category data: ${JSON.stringify(data)}`);
}

// ── Difficulty encoding ─────────────────────────────────

const DIFFICULTY_INDEX: Record<Difficulty, number> = {
  Easy: 0,
  Medium: 1,
  Hard: 2,
  Expert: 3,
};

const INDEX_DIFFICULTY: Difficulty[] = ["Easy", "Medium", "Hard", "Expert"];

function encodeDifficulty(diff: Difficulty): PlutusData {
  return constr(DIFFICULTY_INDEX[diff], []);
}

function decodeDifficulty(data: PlutusData): Difficulty {
  if (isConstr(data)) {
    const diff = INDEX_DIFFICULTY[data.constructor];
    if (diff) return diff;
  }
  throw new Error(`Invalid difficulty data: ${JSON.stringify(data)}`);
}

// ── VerificationType encoding ───────────────────────────

function encodeVerification(v: VerificationType): PlutusData {
  switch (v.type) {
    case "Optimistic":
      return constr(0, [BigInt(v.disputeWindowMs)]);
    case "HumanReview":
      return constr(1, []);
    case "OracleSigned":
      return constr(2, [v.oraclePubKey]);
    case "JudgeAgent":
      return constr(3, []);
  }
}

function decodeVerification(data: PlutusData): VerificationType {
  if (isConstr(data)) {
    switch (data.constructor) {
      case 0:
        return { type: "Optimistic", disputeWindowMs: Number(data.fields[0]) };
      case 1:
        return { type: "HumanReview" };
      case 2:
        return { type: "OracleSigned", oraclePubKey: data.fields[0] as string };
      case 3:
        return { type: "JudgeAgent" };
    }
  }
  throw new Error(`Invalid verification data: ${JSON.stringify(data)}`);
}

// ── BountyStatus encoding ───────────────────────────────

function encodeStatus(s: BountyStatus): PlutusData {
  switch (s.type) {
    case "Open":
      return constr(0, []);
    case "Claimed":
      return constr(1, [s.agent, BigInt(s.claimedAt)]);
    case "WorkSubmitted":
      return constr(2, [s.resultIpfs, BigInt(s.submittedAt)]);
    case "Completed":
      return constr(3, [s.resultIpfs, BigInt(s.completedAt)]);
    case "Disputed":
      return constr(4, []);
    case "Refunded":
      return constr(5, []);
  }
}

function decodeStatus(data: PlutusData): BountyStatus {
  if (isConstr(data)) {
    switch (data.constructor) {
      case 0:
        return { type: "Open" };
      case 1:
        return {
          type: "Claimed",
          agent: data.fields[0] as string,
          claimedAt: Number(data.fields[1]),
        };
      case 2:
        return {
          type: "WorkSubmitted",
          resultIpfs: data.fields[0] as string,
          submittedAt: Number(data.fields[1]),
        };
      case 3:
        return {
          type: "Completed",
          resultIpfs: data.fields[0] as string,
          completedAt: Number(data.fields[1]),
        };
      case 4:
        return { type: "Disputed" };
      case 5:
        return { type: "Refunded" };
    }
  }
  throw new Error(`Invalid status data: ${JSON.stringify(data)}`);
}

// ── AssetId encoding ────────────────────────────────────

function encodeAssetId(a: { policyId: string; assetName: string }): PlutusData {
  return constr(0, [a.policyId, a.assetName]);
}

function decodeAssetId(data: PlutusData): { policyId: string; assetName: string } {
  if (isConstr(data)) {
    return {
      policyId: data.fields[0] as string,
      assetName: data.fields[1] as string,
    };
  }
  throw new Error(`Invalid asset ID data`);
}

// ── BountyDatum ─────────────────────────────────────────

export function encodeBountyDatum(datum: BountyDatum): PlutusData {
  return constr(0, [
    datum.bountyId,
    BigInt(datum.version),
    datum.poster,
    datum.title,
    datum.descriptionIpfs,
    encodeCategory(datum.category),
    encodeDifficulty(datum.difficulty),
    datum.rewardLovelace,
    optionToData(datum.rewardToken, (v) => encodeAssetId(v as { policyId: string; assetName: string })),
    datum.depositLovelace,
    BigInt(datum.deadline),
    BigInt(datum.claimWindowMs),
    encodeVerification(datum.verification),
    BigInt(datum.disputeWindowMs),
    optionToData(datum.resultSchemaIpfs),
    optionToData(datum.allowedAgents, (v) => (v as string[])),
    optionToData(datum.minReputationScore, (v) => BigInt(v as number)),
    encodeStatus(datum.status),
    BigInt(datum.createdAt),
    datum.metadataHash,
  ]);
}

export function decodeBountyDatum(data: PlutusData): BountyDatum {
  if (!isConstr(data)) {
    throw new Error("Invalid BountyDatum data");
  }
  const f = data.fields;
  return {
    bountyId: f[0] as string,
    version: Number(f[1]),
    poster: f[2] as string,
    title: f[3] as string,
    descriptionIpfs: f[4] as string,
    category: decodeCategory(f[5]!),
    difficulty: decodeDifficulty(f[6]!),
    rewardLovelace: BigInt(f[7] as bigint),
    rewardToken: dataToOption(f[8]!, decodeAssetId),
    depositLovelace: BigInt(f[9] as bigint),
    deadline: Number(f[10]),
    claimWindowMs: Number(f[11]),
    verification: decodeVerification(f[12]!),
    disputeWindowMs: Number(f[13]),
    resultSchemaIpfs: dataToOption(f[14]!, (d) => d as string),
    allowedAgents: dataToOption(f[15]!, (d) => d as string[]),
    minReputationScore: dataToOption(f[16]!, (d) => Number(d)),
    status: decodeStatus(f[17]!),
    createdAt: Number(f[18]),
    metadataHash: f[19] as string,
  };
}

// ── WalletPolicyDatum ───────────────────────────────────

export function encodeWalletPolicyDatum(datum: WalletPolicyDatum): PlutusData {
  return constr(0, [
    datum.ownerPkh,
    datum.agentPkh,
    datum.dailyLimitLovelace,
    datum.perTxLimitLovelace,
    datum.whitelistedScripts,
    datum.whitelistedAddresses,
    datum.requireOwnerAbove,
    datum.allowedTokens.map(encodeAssetId),
    datum.maxFeeLovelace,
    optionToData(datum.pauseUntil, (v) => BigInt(v as number)),
    optionToData(
      datum.multisigThreshold,
      (v) => {
        const ms = v as { n: number; keys: string[] };
        return constr(0, [BigInt(ms.n), ms.keys]);
      },
    ),
    BigInt(datum.nonce),
    BigInt(datum.policyVersion),
  ]);
}

export function decodeWalletPolicyDatum(data: PlutusData): WalletPolicyDatum {
  if (!isConstr(data)) {
    throw new Error("Invalid WalletPolicyDatum data");
  }
  const f = data.fields;
  return {
    ownerPkh: f[0] as string,
    agentPkh: f[1] as string,
    dailyLimitLovelace: BigInt(f[2] as bigint),
    perTxLimitLovelace: BigInt(f[3] as bigint),
    whitelistedScripts: f[4] as string[],
    whitelistedAddresses: f[5] as string[],
    requireOwnerAbove: BigInt(f[6] as bigint),
    allowedTokens: (f[7] as PlutusData[]).map(decodeAssetId),
    maxFeeLovelace: BigInt(f[8] as bigint),
    pauseUntil: dataToOption(f[9]!, (d) => Number(d)),
    multisigThreshold: dataToOption(f[10]!, (d) => {
      if (isConstr(d)) {
        return { n: Number(d.fields[0]), keys: d.fields[1] as string[] };
      }
      throw new Error("Invalid multisig data");
    }),
    nonce: Number(f[11]),
    policyVersion: Number(f[12]),
  };
}

// ── ReputationDatum ─────────────────────────────────────

export function encodeReputationDatum(datum: ReputationDatum): PlutusData {
  return constr(0, [
    datum.agentPkh,
    BigInt(datum.totalCompleted),
    datum.totalEarnedLovelace,
    BigInt(datum.totalDisputed),
    BigInt(datum.successRateBps),
    BigInt(datum.avgCompletionMs),
    datum.categoryScores.map((cs) => [cs.categoryId, BigInt(cs.scoreBps)]),
    BigInt(datum.lastActive),
    datum.badgeTokens,
    BigInt(datum.registeredAt),
  ]);
}

export function decodeReputationDatum(data: PlutusData): ReputationDatum {
  if (!isConstr(data)) {
    throw new Error("Invalid ReputationDatum data");
  }
  const f = data.fields;
  return {
    agentPkh: f[0] as string,
    totalCompleted: Number(f[1]),
    totalEarnedLovelace: BigInt(f[2] as bigint),
    totalDisputed: Number(f[3]),
    successRateBps: Number(f[4]),
    avgCompletionMs: Number(f[5]),
    categoryScores: (f[6] as [string, bigint][]).map(([id, score]) => ({
      categoryId: id,
      scoreBps: Number(score),
    })),
    lastActive: Number(f[7]),
    badgeTokens: f[8] as string[],
    registeredAt: Number(f[9]),
  };
}

// ── BountyRedeemer encoding ─────────────────────────────

export function encodeBountyRedeemer(redeemer: BountyRedeemer): PlutusData {
  switch (redeemer.type) {
    case "PostBounty":
      return constr(0, []);
    case "ClaimBounty":
      return constr(1, [redeemer.agentAddress]);
    case "ExtendClaim":
      return constr(2, [BigInt(redeemer.newClaimWindowMs)]);
    case "SubmitWork":
      return constr(3, [redeemer.resultIpfs]);
    case "VerifyAndPay":
      return constr(4, []);
    case "DisputeWork":
      return constr(5, []);
    case "ResolveDispute":
      return constr(6, [redeemer.winner]);
    case "RefundPoster":
      return constr(7, []);
    case "CancelBounty":
      return constr(8, []);
    case "UpdateBounty":
      return constr(9, [redeemer.newTitle, redeemer.newDescriptionIpfs]);
  }
}
